### DISCUSSION TITLE
GdUnit4 V4.1.0 feedback

### DISCUSSION THREAD

**Discussion by MikeSchulze** [MikeSchulze/gdUnit4#157] [General]
**You are invited to give me feedback on the GdUnit4.**

---
Please describe what you like and what you don't ;)
I ask for a factual discussion and constructive criticism here.

---
If you found bugs, please report them [here](https://github.com/MikeSchulze/gdUnit4/issues/new?assignees=MikeSchulze&labels=bug&template=bug_report.md&title=)
Also feel free to create [feature requests](https://github.com/MikeSchulze/gdUnit4/issues/new?assignees=MikeSchulze&labels=enhancement&template=feature_request.md&title=).

---
Thanks in advance
Mike

**Comment by devpitOne** [comment]
Just tried this out and successfully implemented the tutorial. Very happy to see this level of integrated unit test support for Godot.

I was thinking of switching over to C#, so the C# support feature is of great interest to me.

**Reply by MikeSchulze** [reply]
> Just tried this out and successfully implemented the tutorial. Very happy to see this level of integrated unit test support for Godot.
> 
> I was thinking of switching over to C#, so the C# support feature is of great interest to me.

Thanks for the feedback, C# support will be added again with v4.2.0.

**Comment by SdgGames** [comment]
I just got started with GDUnit4, and I'm loving it! One question - is there a concise way to run ALL tests? My project isn't that big, but it feels like I'm always running a subset of tests whenever I hit the "run tests" button. That just runs tests on the current file, right? Apologies if I missed something obvious, I just don't want tests to start failing silently if I'm always running tests against just the class I'm writing.

**Reply by MikeSchulze** [reply]
Thanks for the nice feedback.

You can easily run all tests by selecting a folder and using the context menu.
![image](https://github.com/MikeSchulze/gdUnit4/assets/347037/3042380a-7a3e-4d98-83fa-536e4c827ff9)
I checked the documentation and it looks like I forgot to mention this option.

You can also enable the "Run Overall" button in the UI settings.
Then you get an additional button in the inspector to run all tests at once.
![image](https://github.com/MikeSchulze/gdUnit4/assets/347037/8eb6b6f8-1b3e-498a-aa95-b51bb3d5b918)

**Reply by MikeSchulze** [reply]
The documentation is now update and has a new FAQ entry
https://mikeschulze.github.io/gdUnit4/faq/run-tests/

**Comment by andrew-wilkes** [comment]
I just want to say that I like it a lot. I am re-coding one of my projects and wanted to implement better testing. So far I have implemented some unit tests on class scripts. My project code is here: https://github.com/andrew-wilkes/logic-simulator-2/tree/main

**Comment by megonemad1** [comment]
Hi, is there a way to kind of do parameterized tests but based on the result of a function call against the scene runner?

I have a scene that defines states it can take and in an attempt to not repeat myself I want to call the scenes .get_state_list and then run the scene as a scene runner with each and the test UI properly distinguish between each option as a test.

**Reply by MikeSchulze** [reply]
Dynamic data sets are not supported. Parameterized test uses static predefined data sets.

To check scene state changes, you can use signals to await for a state. 
Or wait for a function state
```
func test_wait_for_scene_state() -> void:
	var scene := runner.scene()
	await assert_func(scene, "get_state").wait_until(1000).is_equal(10)
```

**Reply by megonemad1** [reply]
That's not quite the usecase I had, but good to know it doesn't support dynamic tests based on non static test cases

**Reply by polzon** [reply]
A big blocker for me is the lack of running headless support. 

I'm not sure if that's a Godot limitation since I'm pretty sure the GUT addon also lacks support for it. I think it could open up a lot more possible tooling if this was possible. 

I do see the cli tool has an override to enable it, but didn't have much success with it. Also for some reason it locks up VSCode when I had an agent try to run headless tests.

I was just curious on the state of that since the extension just says it's not allowed. I haven't found any discussion about this so I was wondering if there are plans for official support?

**Reply by MikeSchulze** [reply]
> A big blocker for me is the lack of running headless support.
> 
> I'm not sure if that's a Godot limitation since I'm pretty sure the GUT addon also lacks support for it. I think it could open up a lot more possible tooling if this was possible.
> 
> I do see the cli tool has an override to enable it, but didn't have much success with it. Also for some reason it locks up VSCode when I had an agent try to run headless tests.
> 
> I was just curious on the state of that since the extension just says it's not allowed. I haven't found any discussion about this so I was wondering if there are plans for official support?

Hi @polzon 

However, you can run the command-line tool in headless mode; to do so, you must use the parameters 
`--headless --ignoreHeadlessMode`.
Headless mode is disabled by default to prevent test errors that inevitably occur during UI tests, since Godot cannot process UI input in headless mode.
So if you don’t have any UI tests, you can run the tests in headless mode.
